java/src/main/java/ai/rapids/cudf/Rmm.java

```diff
@@ -171,32 +235,76 @@ public static void shutdown(long forceGCInterval, long maxWaitTime, TimeUnit uni
     if (MemoryCleaner.bestEffortHasRmmBlockers()) {
       throw new RmmException("Could not shut down RMM there appear to be outstanding allocations");
     }
-    shutdownInternal();
+    if (initialized) {
+      shutdownInternal();
+      initialized = false;
+    }
   }
 
-  private native static void shutdownInternal() throws RmmException;
+  private static native void shutdownInternal() throws RmmException;
 
   /**
-   * ---------------------------------------------------------------------------*
-   * Allocate memory and return a pointer to device memory.
-   * If initialization has not occurred then RMM will be implicitly initialized
-   * in CUDA default mode.
-   * <p>
-   * Mapping: RMM_ALLOC in rmm.h.
+   * Small class used to ensure that we follow the rules of RMM allocation
+   * and keep the size of the allocation around so we can free it without
+   * any issues.  This keeps the stream around too, but that may change
+   * in the future when we have better stream support, and the stream you
+   * allocate something on is not always the one you want to synchronize against
+   * when you free it.
+   */
+  public static class RmmBuff implements AutoCloseable {
```

Comment:
Should this be public? I'm thinking we only want to expose `DeviceMemoryBuffer`, or is there a use-case for apps to access `RmmBuff` directly that cannot be served via `DeviceMemoryBuffer`?  I wonder if there is a real need for `RmmBuff` or if these details should be moved into `DeviceMemoryBuffer` directly.

If this remains, I think we should spring for the two extra letters for consistency with `MemoryBuffer`, `DeviceMemoryBuffer`, and `HostMemoryBuffer`.

Reply:
I made it public because the original RMM API was public.  I am happy to make it package private.

Reply:
Is there a good reason for it to be separate from DeviceMemoryBuffer?  The more I think about it, the more it feels like RMM should just return that directly.